### PR TITLE
Add equation element parsing grammar

### PR DESCRIPTION
**Pull request by brynpickering**
Partial fix for issue(s) #360, #100 to be able to process 5 types of equation elements for our new constraints (overall plan detailed in #100):

* Numbers including "inf"
* Unindexed parameters  / decision variables of the form "foo"
* Indexed parameters  / decision variables of the form "foo[foo, bar]"
* _Components_ of equations, which will be dealt with later and we're simply catching for now, of the form "$foo"
* Helper functions, which can take any of the above - as well as themselves - as inputs, of the form "foo(*args, **kwargs)".

Reviewer checklist:

- [x] Test(s) added to cover contribution
- [ ] Documentation updated
- [ ] Changelog updated
- [x] Coverage maintained or improved

### CODE REVIEW

**Comment by brynpickering** [comment]
As I have commented in-line, there is one design choice that I have made that could be changed. Generally, I am only checking the parsed data for consistency with the Calliope model on evaluation. 

E.g., I could check the helper function on parsing by only allowing string values from the list of helper function dictionary keys. Instead, I allow any helper function string, but only on evaluation (actually converting the string into a pointer to the function) do I check that it is a valid helper function. In this way, I have separated string parsing operations from anything related to user/calliope internal data until the last moment. I don't know that it is the best option, though! Maybe it makes more sense to have things fail directly on parsing rather than successfully parsing everything and then failing on evaluation...

**Comment by sjpfenninger** [comment]
> ...
> The exception to this rule is the indexed variable/parameter parsing, which uses the known set iterators to limit the strings that can be found between the `[]` brackets. I would move this to the evaluation step if I were to align everything.

I think aligning this by moving to evaluation stage makes sense.

> I don't know that it is the best option, though! Maybe it makes more sense to have things fail directly on parsing rather than successfully parsing everything and then failing on evaluation...

I see no issue with this. It keeps the parsing code somewhat cleaner, and we think that parsing will always be fast enough that not failing on parsing is not an issue. So I would keep it as is.

**Comment by brynpickering** [comment]
The other side of the coin is that because parsing is so quick, it might be best to fail there. Evaluation might end up being (relatively) slow as it's when the interface with pyomo/gurobi is instantiated, so catching errors then could be more annoying.

Things that we would need to check either at evaluation or during parsing:
1. set iterators are given in `foreach` (or, later, as an `index_item`) 
2. sets are defined in `model_data.dims`
2. components are defined in `components`
3. params are in `model_data.data_vars`
4. variables are in the optimisation model object
5. indexed params/variables are indexed over all relevant dimensions
6. helper functions are part of a pre-defined dictionary

**Review by sjpfenninger — DISMISSED** [review]

File: calliope/backend/equation_parser.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+import pyparsing as pp
+
+
+pp.ParserElement.enablePackrat()
+
+COMPONENT_CLASSIFIER = "$"
+
+
+class EvalFunction:
```
[the remark this answers is not in the document]
Comment (brynpickering) [reply calliope/backend/equation_parser.py:38]: Done.

**Pull request merged**